Fix/tooltip refresh on change context object

IIsaku opened the pull request:
This fixes #9 by attaching the tooltip in the correct way to the target element. This fix also includes improvements for npm scripts 'start' and 'build'.

Commit:
Replace Gulp with Webpack

- Use NPM for dependency management
- Convert Mx test project to Mx6.10.18
- Configure webpack

IIsaku committed:
Improve webpack config

- Remove unused npm dependencies
- Improve npm scripts
- Reduce webpack config files
- Remove local config
- Set xml templating only for widget version
- Adjust license copyright details

IIsaku committed:
Remove support below mx 6 & add prettier

IIsaku committed:
Adjust .gitignore

IIsaku committed:
Add test projects

- Change test folder structure
- Add test projects for mx 6 latest, mx 7 first & mx 8 first
- Reconfigure prettier
- Improve webpack configuration

IIsaku committed:
Adjust test projects to improve manual testing

IIsaku committed:
Update npm package info

IIsaku committed:
Improve readme

IIsaku committed:
Add code blocks for terminal commands in README

IIsaku committed:
Fix issue with mx 8 test project

IIsaku committed:
Fix tooltip attachment to target element

- Tooltip attaches once to target element
- The tooltip text will be evaluated with a function
- Improve test projects

IIsaku committed:
Improve npm scripts

- The start & build script will now update source files in the deployment folder of the test projects
- Simplify webpack configuration by removing a plugin dependency

diego-antonelli commented on .prettierrc:
```diff
@@ -0,0 +1,11 @@
+      }
+    }
+  ]
+}
```
Please try to comply with New Mendix pattern:
```
{
  "trailingComma": "none",
  "useTabs": false,
  "tabWidth": 4,
  "semi": true,
  "singleQuote": false,
  "printWidth": 120,
  "bracketSpacing": true,
  "jsxBracketSameLine": false,
  "arrowParens": "avoid",
  "proseWrap": "always"
}
```

IIsaku replied:
Done, but what about the overrides for json files? Andries told me json files should have tabwidth 2.

Andries-Smit replied:
Just keep as is

diego-antonelli commented on test/mx-8/theme/LICENSE:
```diff
@@ -0,0 +1,21 @@
+MIT License
```
Do we need the theme and widgets committed to this test folder?

IIsaku replied:
Yes, because otherwise the styling of the app will break due to a missing atlas UI dependency. The project also depends on other widgets, so the widgets folder needs to be included as well. However, the BootstrapTooltip widget isn't committed.

IIsaku committed:
Change prettier config & remove unused npm packages

IIsaku committed:
Improve tooltip message with attribute source & improve hover trigger on mobile

- Changes in the tooltip message attribute will now be displayed in the tooltip text due to subscription to the attribute. Beforehand, the tooltip message would only change when the context object changed.
- When the trigger for a tooltip is set to hover, this will be converted to a click trigger on mobile devices.
- Remove unnecessary theme data in mx 8 test project.

IIsaku committed:
Ignore native and sass folder in mx 8 test project

IIsaku commented:
This pull request now also improves a tooltip message with attribute source and hover triggering on mobile. It fixes #11 #8.

Andries-Smit commented on .prettierignore:
```diff
@@ -0,0 +1,7 @@
+*.xml
+.gitignore
+.prettierignore
+.DS_Store
```
missing newline

Andries-Smit commented on .gitignore:
```diff
@@ -1,20 +1,25 @@
 
-node_modules/
+.idea/
+.vscode/
```
suggest to add this to you global ~/.gitignore too

Andries-Smit commented on README.md:
```diff
@@ -1,31 +1,95 @@
+
+-   Locate the widget as close to the target element as possible. (Preferably next to it)
+-   When an element has a tooltip attached to it and needs to be conditionally visible, put both element and tooltip
+    together in a container and apply the visibility conditions to the container instead of the element.
```
Add screen shot of the suggestion

Andries-Smit commented on README.md:
```diff
@@ -1,31 +1,95 @@
 
-## Properties
+-   Based on Bootstrap's Tooltip.js
+
```
Your tip might be described as a limitation....
- Not recommended to use in Mendix 8? Discuss with PO

Andries-Smit commented on src/BootstrapTooltip.js:
```diff
@@ -0,0 +1,142 @@
+import declare from "dojo/_base/declare";
```
Suggestion, for next time, to commit the move separately so you can diff

Andries-Smit commented on src/BootstrapTooltip.js:
```diff
@@ -0,0 +1,142 @@
+
+    _tooltipText: "No custom text specified for this tooltip",
+    _tooltipTrigger: null,
+
```
remove this empty line

Andries-Smit commented on src/BootstrapTooltip.js:
```diff
@@ -0,0 +1,142 @@
+            this._tooltipTrigger = "focus";
+        } else if (this.tooltipMode === "click") {
+            this._tooltipTrigger = "click";
+        }
```
could rewrite
``` ts
this._tooltipTrigger = this.tooltipMode === "hover" ? "focus hover" : this.tooltipMode;
```

Andries-Smit commented on src/BootstrapTooltip.js:
```diff
@@ -0,0 +1,142 @@
+            this._execMf(
+                this.tooltipMessageMicroflow,
+                obj ? obj.getGuid() : null,
+                lang.hitch(this, function(string) {
```
be consistent, use arrow functions everywhere or no where....

Andries-Smit replied:
not recommended to use `string` as variable name

Andries-Smit commented on src/BootstrapTooltip.js:
```diff
@@ -0,0 +1,142 @@
+                );
+            }
+
+            //if the element is a label+input combination, find the input element.
```
prettier should add space after double slash `// blabalb`

IIsaku replied:
Prettier doesn't support that: https://github.com/prettier/prettier/issues/5755#issuecomment-454276995

Andries-Smit commented on test/mx-8/theme/README.md:
```diff
@@ -0,0 +1,83 @@
+# Mendix Atlas UI
```
add to ignore

Andries-Smit commented on test/mx-8/theme/styles.js:
```diff
@@ -0,0 +1,12 @@
+import * as main from './styles/native/main';
```
ignore file

Andries-Smit commented on webpack.config.js:
```diff
@@ -0,0 +1,179 @@
+module.exports = {
+    mode: isDev ? MODES.DEV : MODES.PROD,
+    target: "web",
+    devtool: isDev ? "eval-source-map" : false,
```
in future we will use `source-map` bit slower, but will show the same as the  source 
https://webpack.js.org/configuration/devtool/

Andries-Smit commented on webpack.config.js:
```diff
@@ -0,0 +1,179 @@
+        libraryTarget: "amd"
+    },
+    optimization: {
+        minimizer: [
```
Discuss for mono repo, if dojo widgets should be minified as source maps are not supported for bundled widgets.js

Andries-Smit commented on webpack.config.js:
```diff
@@ -0,0 +1,179 @@
+}
+
+function initFileManagerPlugin(plugins) {
+    const mxTestProjectVersions = [6, 7, 8];
```
one way of doing, it tough for the mono repo, we list the location for the test project in the package.json

Andries-Smit commented on src/BootstrapTooltip.js:
```diff
@@ -0,0 +1,142 @@
+                html: this.tooltipRenderHTML
+            });
+
+            !this._tooltipInitialized;
```
dont like this....

Andries-Smit commented on src/BootstrapTooltip.js:
```diff
@@ -0,0 +1,142 @@
+    },
+
+    _getTrigger: function() {
+        return this._tooltipTrigger === "hover" && this.isMobileDevice() ? "click" : this._tooltipTrigger;
```
Why do this? add comment

IIsaku committed:
Introduce changes from pull request feedback

IIsaku committed:
Add removed files in mx-8 test project to gitignore

IIsaku committed:
Bump version to 3.2.0

IIsaku commented on src/BootstrapTooltipContext.js:
```diff
@@ -73,6 +61,5 @@ export default declare("BootstrapTooltip.widget.BootstrapTooltipContext", [_boot
             this._tooltipText = this.tooltipMessageString;
         }
-        this._initializeTooltip();
     }
```
why removed, does it still work?

IIsaku commented on src/BootstrapTooltip.js:
```diff
@@ -120,14 +107,12 @@ export default declare("BootstrapTooltip.widget.BootstrapTooltip", [_WidgetBase,
-                console.log(this.id + "._execMf error: " + error.description);
-            })
+            },
+            error: error => console.log(this.id + "._execMf error: " + error.description)
```
error.descriptions does not excists

IIsaku replied:
Error will now be exposed to console.

IIsaku commented on src/BootstrapTooltip.js:
```diff
@@ -90,13 +78,12 @@ export default declare("BootstrapTooltip.widget.BootstrapTooltip", [_WidgetBase,
         this._executeCallback(cb, "_initializeTooltip");
     },
 
+    // Translate hover trigger to click trigger on mobile devices
```
Comment why, not what. What is described in the code

IIsaku replied:
Done, why explained

IIsaku commented on src/BootstrapTooltip.js:
```diff
@@ -55,33 +42,34 @@ export default declare("BootstrapTooltip.widget.BootstrapTooltip", [_WidgetBase,
+                    : $targetElement.find("input");
+        }
+
+        if ($targetElement.length > 0) {
```
What else?

IIsaku replied:
Done, else: message that input in formgroup could not be found

IIsaku commented on README.md:
```diff
@@ -37,7 +37,10 @@ Extra for Bootstrap Tooltip (Context):
 -   When an element has a tooltip attached to it and needs to be conditionally visible, put both element and tooltip
-    together in a container and apply the visibility conditions to the container instead of the element.
+    together in a container and apply the visibility conditions to the container instead of the element (see the image
+    below).
```
prevent newlines (maybe prettier, or IDE?)

IIsaku replied:
Done

IIsaku committed:
Apply feedback improvements to readme and code

IIsaku committed:
Add test cases cond vis

IIsaku committed:
Remove unused widgets from mx 8 test project

IIsaku committed:
Ignore all widgets in test
